File: ios/App/App/Plugins/NavigationPlugin.swift
import Foundation
import Capacitor

/**
 * NavigationPlugin (iOS)
 *
 * Ponte entre o front-end TypeScript e o Google Maps Navigation SDK
 * no iOS. Apresenta um NavigationViewController por cima do WebView,
 * exibindo mapa turn-by-turn com voz nativa — sem sair do app.
 *
 * Registro em AppDelegate.swift:
 *   CAPBridge.registerPlugin(NavigationPlugin.self)
 *
 * Ou via capacitor.config.json:
 *   { "plugins": { "CapacitorNavigation": {} } }
 */
@objc(NavigationPlugin)
public class NavigationPlugin: CAPPlugin, CAPBridgedPlugin {

    // ─── Identificação do plugin (deve coincidir com registerPlugin no TS) ────
    public let identifier   = "NavigationPlugin"
    public let jsName       = "CapacitorNavigation"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "isAvailable",     returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "startNavigation", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "stopNavigation",  returnType: CAPPluginReturnPromise),
    ]

    // Referência ao ViewController de navegação ativo
    private var navController: NavigationViewController?

    // ─── isAvailable ─────────────────────────────────────────────────────────

    @objc func isAvailable(_ call: CAPPluginCall) {
        // Verifica se a classe principal do Navigation SDK existe no bundle
        let available = NSClassFromString("GMSNavigationServices") != nil
        call.resolve(["available": available])
    }

    // ─── startNavigation ─────────────────────────────────────────────────────

    @objc func startNavigation(_ call: CAPPluginCall) {
        guard
            let lat   = call.getDouble("lat"),
            let lng   = call.getDouble("lng")
        else {
            call.reject("Parâmetros 'lat' e 'lng' são obrigatórios.")
            return
        }
        let label = call.getString("label") ?? "Destino"

        DispatchQueue.main.async { [weak self] in
            guard let self = self else { return }

            let vc = NavigationViewController(
                latitude:  lat,
                longitude: lng,
                label:     label
            )

            // Callback de conclusão: resolve a promise JS quando o VC fechar
            vc.onFinished = { [weak self] error in
                self?.navController = nil
                if let error = error {
                    call.reject(error)
                } else {
                    call.resolve(["initialized": true])
                }
            }

            self.navController = vc

            // Apresenta sobre o WebView do Capacitor em tela cheia
            vc.modalPresentationStyle = .fullScreen
            self.bridge?.viewController?.present(vc, animated: true)
        }
    }

    // ─── stopNavigation ──────────────────────────────────────────────────────

    @objc func stopNavigation(_ call: CAPPluginCall) {
        DispatchQueue.main.async { [weak self] in
            self?.navController?.dismiss(animated: true) {
                self?.navController = nil
                call.resolve()
            }
        }
    }
}

File: android/app/src/main/java/app/uppi/mobile/plugins/NavigationPlugin.kt
package app.uppi.mobile.plugins

import android.content.Intent
import android.util.Log
import com.getcapacitor.JSObject
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.annotation.CapacitorPlugin

/**
 * CapacitorNavigationPlugin
 *
 * Ponte entre o front-end TypeScript e o Google Maps Navigation SDK nativo
 * no Android. Abre a NavigationActivity (que embute o SupportNavigationFragment)
 * e retorna o resultado para o JS via PluginCall.
 *
 * Registro em MainAcitivity.kt:
 *   add(NavigationPlugin::class.java)
 */
@CapacitorPlugin(name = "CapacitorNavigation")
class NavigationPlugin : Plugin() {

    companion object {
        private const val TAG = "CapacitorNavigation"

        // Request code para identificar o retorno da NavigationActivity
        const val REQUEST_NAVIGATION = 9001
    }

    // Guarda a call ativa para resolver depois que a Activity retornar
    private var pendingCall: PluginCall? = null

    // ─── isAvailable ─────────────────────────────────────────────────────────

    @PluginMethod
    fun isAvailable(call: PluginCall) {
        // O Navigation SDK está disponível se a dependência foi incluída no build.gradle.
        // Verificamos dinamicamente se a classe principal do SDK existe no classpath.
        val available = try {
            Class.forName("com.google.android.libraries.navigation.NavigationApi")
            true
        } catch (e: ClassNotFoundException) {
            Log.w(TAG, "Navigation SDK não encontrado no classpath: ${e.message}")
            false
        }

        val result = JSObject()
        result.put("available", available)
        call.resolve(result)
    }

    // ─── startNavigation ─────────────────────────────────────────────────────

    @PluginMethod
    fun startNavigation(call: PluginCall) {
        val lat   = call.getDouble("lat")
        val lng   = call.getDouble("lng")
        val label = call.getString("label") ?: "Destino"

        if (lat == null || lng == null) {
            call.reject("Parâmetros 'lat' e 'lng' são obrigatórios.")
            return
        }

        // Salva a call para resolver quando a Activity terminar
        pendingCall = call

        val intent = Intent(activity, NavigationActivity::class.java).apply {
            putExtra(NavigationActivity.EXTRA_LAT,   lat)
            putExtra(NavigationActivity.EXTRA_LNG,   lng)
            putExtra(NavigationActivity.EXTRA_LABEL, label)
        }

        Log.d(TAG, "Iniciando NavigationActivity → lat=$lat lng=$lng label=$label")
        startActivityForResult(call, intent, REQUEST_NAVIGATION)
    }

    // ─── stopNavigation ──────────────────────────────────────────────────────

    @PluginMethod
    fun stopNavigation(call: PluginCall) {
        // Encerra a NavigationActivity se estiver em foreground
        activity.runOnUiThread {
            // Envia um broadcast para a Activity se fechar
            val intent = Intent(NavigationActivity.ACTION_STOP_NAVIGATION)
            activity.sendBroadcast(intent)
        }
        call.resolve()
    }

    // ─── Callback quando a NavigationActivity retorna ────────────────────────

    override fun handleOnActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.handleOnActivityResult(requestCode, resultCode, data)

        if (requestCode != REQUEST_NAVIGATION) return

        val savedCall = pendingCall ?: return
        pendingCall = null

        val result = JSObject()
        result.put("initialized", true)

        Log.d(TAG, "NavigationActivity retornou, resultCode=$resultCode")
        savedCall.resolve(result)
    }
}
